tests/h/accounts/schemas_test.py

```diff
@@ -377,27 +377,138 @@ def test_it_is_invalid_if_password_wrong(self,
-                                             factories,
-                                             pyramid_csrf_request,
-                                             user_model):
-        # There isn't already an account with the email address that we're
-        # trying to change to.
-        user_model.get_by_email.return_value = None
+    # The user's password.
+    PASSWORD = 'flibble'
+
+    def test_it_returns_the_new_email_when_valid(self, schema):
+        appstruct = schema.deserialize({
+            'email': 'foo@bar.com',
+            'password': self.PASSWORD,
+        })
+
+        assert appstruct['email'] == 'foo@bar.com'
+
+    def test_it_is_valid_if_email_same_as_users_existing_email(self,
+                                                               schema,
+                                                               user,
+                                                               user_model,
+                                                               pyramid_config):
+        """
+        It is valid if the new email is the same as the user's existing one.
+
+        Trying to change your email to what your email already is should not
+        return an error.
+
+        """
+        user_model.get_by_email.return_value = Mock(spec_set=['userid'],
+                                                    userid=user.userid)
+        pyramid_config.testing_securitypolicy(user.userid)
+
+        schema.deserialize({'email': user.email, 'password': self.PASSWORD})
+
+    def test_it_is_invalid_if_csrf_token_missing(self,
+                                                 pyramid_request,
+                                                 schema):
+        del pyramid_request.headers['X-CSRF-Token']
 
-        pyramid_csrf_request.authenticated_user = factories.User()
-        db_session.add(pyramid_csrf_request.authenticated_user)
+        with pytest.raises(BadCSRFToken):
+            schema.deserialize({
+                'email': 'foo@bar.com',
+                'password': self.PASSWORD,
+            })
 
-        schema = schemas.EmailChangeSchema().bind(request=pyramid_csrf_request)
+    def test_it_is_invalid_if_csrf_token_wrong(self, pyramid_request, schema):
+        pyramid_request.headers['X-CSRF-Token'] = 'WRONG'
 
+        with pytest.raises(BadCSRFToken):
+            schema.deserialize({
+                'email': 'foo@bar.com',
+                'password': self.PASSWORD,
+            })
+
+    def test_it_is_invalid_if_password_wrong(self, schema):
         with pytest.raises(colander.Invalid) as exc:
             schema.deserialize({
                 'email': 'foo@bar.com',
-                'password': 'flibble'  # Not the correct password!
+                'password': 'WRONG'  # Not the correct password!
             })
 
-        assert 'password' in exc.value.asdict()
+        assert exc.value.asdict() == {
+            'password': 'Incorrect password. Please try again.'}
```

Comment:
I usually don't mind these test cases. It is much better to include the error message than to not be exactly sure which error happened.